diff -rupN '--exclude=.*' ../SDCC/Makefile ./Makefile
--- ../SDCC/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./Makefile	2013-02-22 13:29:45.000000000 +0200
@@ -0,0 +1,23 @@
+
+all: tool install demo
+
+tool:
+	$(MAKE) -C tool all
+
+install:
+	$(MAKE) -C tool install
+
+demo: 	
+	@echo
+	@echo "***************************************";
+	@echo "***      Example build AR168M";
+	@echo "*** To build another configuration:";
+	@echo "*** look at the full command below ;)";
+	@echo "***************************************";
+	cd src; chmod +x mk; ./mk ar168m sip us
+
+.PHONY: tool
+
+clean:
+	$(MAKE) -C tool clean
+	$(MAKE) -C src clean
diff -rupN '--exclude=.*' ../SDCC/ar1688_linux.README ./ar1688_linux.README
--- ../SDCC/ar1688_linux.README	1970-01-01 02:00:00.000000000 +0200
+++ ./ar1688_linux.README	2013-02-22 13:46:09.000000000 +0200
@@ -0,0 +1,146 @@
+Compile ar1688 API on linux port by AlexVangelov
+------------------------------------------------
+1. Download and extract http://palmmicro.com/ar1688/software/snapshot/AR1688_API_061002.zip
+2. copy AR1688_API_061002_linux.patch to SDCC folder
+4. cd SDCC
+5. make
+------------------------------------------------
+* To compile only tools: 
+   make tool install
+* To build specific configuration go to SDCC/src and execute:
+   ./mk [board] [protocol] [language] [oem]
+*** Original project URL: http://palmmicro.com/ar1688
+
+
+BUILDLOG:
+vangelov@vangelov ~/SDCC $ patch -p1 < AR1688_API_061002_linux.patch 
+patching file Makefile
+patching file include/Ne2000.h
+patching file include/ne2000.h
+patching file include/version.h
+patching file src/makefile
+patching file src/menu.c
+patching file src/mk
+patching file src/time.c
+patching file src/ui_str.c
+patching file tool/Makefile
+patching file tool/common/cmd_default.cpp
+patching file tool/common/common.cpp
+patching file tool/common/mfc2std.h
+patching file tool/hex2bin/HexFile.cpp
+patching file tool/hex2bin/HexFile.h
+patching file tool/hex2bin/Makefile
+patching file tool/hex2bin/StdAfx.h
+patching file tool/hex2bin/hex2bin.cpp
+patching file tool/namebin/Makefile
+patching file tool/namebin/StdAfx.h
+patching file tool/namebin/namebin.cpp
+patching file tool/relink/Makefile
+patching file tool/relink/StdAfx.h
+patching file tool/relink/relink.cpp
+patching file tool/reversion/Makefile
+patching file tool/reversion/StdAfx.h
+patching file tool/reversion/reversion.cpp
+vangelov@vangelov ~/SDCC $ make
+make -C tool all
+make[1]: Entering directory `/home/vangelov/SDCC/tool'
+make[2]: Entering directory `/home/vangelov/SDCC/tool/reversion'
+g++ -c -o reversion.o reversion.cpp -I.
+g++ -c -o ../common/cmd_default.o ../common/cmd_default.cpp -I.
+g++ -o reversion reversion.o ../common/cmd_default.o -I.
+make[2]: Leaving directory `/home/vangelov/SDCC/tool/reversion'
+make[2]: Entering directory `/home/vangelov/SDCC/tool/relink'
+g++ -c -o relink.o relink.cpp -I.
+g++ -o relink relink.o ../common/cmd_default.o -I.
+make[2]: Leaving directory `/home/vangelov/SDCC/tool/relink'
+make[2]: Entering directory `/home/vangelov/SDCC/tool/hex2bin'
+g++ -c -o hex2bin.o hex2bin.cpp -I.
+g++ -c -o HexFile.o HexFile.cpp -I.
+g++ -o hex2bin hex2bin.o HexFile.o -I.
+make[2]: Leaving directory `/home/vangelov/SDCC/tool/hex2bin'
+make[2]: Entering directory `/home/vangelov/SDCC/tool/namebin'
+g++ -c -o namebin.o namebin.cpp -I.
+g++ -o namebin namebin.o ../common/cmd_default.o -I.
+make[2]: Leaving directory `/home/vangelov/SDCC/tool/namebin'
+make[1]: Leaving directory `/home/vangelov/SDCC/tool'
+make -C tool install
+make[1]: Entering directory `/home/vangelov/SDCC/tool'
+*** Copy reversion to bin
+*** Copy relink to bin
+*** Copy hex2bin to bin
+*** Copy namebin to bin
+make[1]: Leaving directory `/home/vangelov/SDCC/tool'
+
+***************************************
+***      Example build AR168M
+*** To build another configuration:
+*** look at the full command below ;)
+***************************************
+cd src; chmod +x mk; ./mk ar168m sip us
+Palmmicro reversion utility 0.55.011
+Palmmicro AR1688 relink utility 0.61.002
+make[1]: Entering directory `/home/vangelov/SDCC/src'
+sdcc main.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdcc ce3.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdasz80 -plosgff datatype.rel datatype.s
+sdcc isr_gpio.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdcc ksz8842.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdcc lcd.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdasz80 -plosgff math.rel math.s
+sdcc ne2000.c -mz80 -I../include -c --std-c99 --codeseg CODE0
+sdcc font.c -mz80 -I../include -c --std-c99 --codeseg CODEF
+sdcc buf.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc data.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc heap.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc line.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc misc.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc options.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc rand.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc sipc.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc syslog.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc udpdebug.c -mz80 -I../include -c --std-c99 --codeseg CODE
+sdcc adc_key.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc ar168r.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc dtmf.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc function.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc gpio_key.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc hook.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc ivr.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc key.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc led.c -mz80 -I../include -c --std-c99 --codeseg CODE1 --disable-warning 110
+sdcc serial.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc time.c -mz80 -I../include -c --std-c99 --codeseg CODE1
+sdcc dhcp.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc dns.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc net.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc ppp.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc pppoe.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc sntp.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc stun.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc tftp.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc udprun.c -mz80 -I../include -c --std-c99 --codeseg CODE2
+sdcc http.c -mz80 -I../include -c --std-c99 --codeseg CODE3
+sdcc httpc.c -mz80 -I../include -c --std-c99 --codeseg CODE3
+sdcc ap_innom.c -mz80 -I../include -c --std-c99 --codeseg CODE3
+sdcc call_log.c -mz80 -I../include -c --std-c99 --codeseg CODE4
+sdcc menu.c -mz80 -I../include -c --std-c99 --codeseg CODE4
+sdcc settings.c -mz80 -I../include -c --std-c99 --codeseg CODE4
+sdcc ui_str.c -mz80 -I../include -c --std-c99 --codeseg CODE4
+sdcc iax2.c -mz80 -I../include -c --std-c99 --codeseg CODE5
+sdcc protocol.c -mz80 -I../include -c --std-c99 --codeseg CODE5
+sdcc rtps.c -mz80 -I../include -c --std-c99 --codeseg CODE5
+sdcc sips.c -mz80 -I../include -c --std-c99 --codeseg CODE5
+sdcc auth.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc md5.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc rc4.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc sha1.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc hmac_sha1.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc pbkdf2_sha1.c -mz80 -I../include -c --std-c99 --codeseg CODE6
+sdcc digitmap.c -mz80 -I../include -c --std-c99 --codeseg CODE7
+sdcc rtpr.c -mz80 -I../include -c --std-c99 --codeseg CODE7
+sdcc sipr.c -mz80 -I../include -c --std-c99 --codeseg CODE7
+sdld -n -f linkmain_unix
+../bin/hex2bin main.bin -f0 -a0 -x6 -b32768 -e57344 main.ihx
+make[1]: Leaving directory `/home/vangelov/SDCC/src'
+Palmmicro AR1688 namebin utility 0.59.028
+
diff -rupN '--exclude=.*' ../SDCC/include/Ne2000.h ./include/Ne2000.h
--- ../SDCC/include/Ne2000.h	2012-06-29 12:33:26.000000000 +0300
+++ ./include/Ne2000.h	1970-01-01 02:00:00.000000000 +0200
@@ -1,186 +0,0 @@
-// The header for an Novell 2000 driver.
-
-#define NE2000_RAM_BASE		0x4000
-#define NE2000_RAM_SIZE		0x2000
-
-//    Ne2000 Hardware definitions.
-
-// Offsets from Adapter->IoPAddr of the ports used to access
-// the 8390 NIC registers.
-//
-// The names in parenthesis are the abbreviations by which
-// the registers are referred to in the 8390 data sheet.
-//
-// Some of the offsets appear more than once
-// because they have have relevant page 0 and page 1 values,
-// or they are different registers when read than they are
-// when written. The notation MSB indicates that only the
-// MSB can be set for this register, the LSB is assumed 0.
-
-#define NIC_COMMAND         0x0     // (CR)
-#define NIC_PAGE_START      0x1     // (PSTART)   MSB, write-only
-#define NIC_PHYS_ADDR       0x1     // (PAR0)     page 1
-#define NIC_9346CR			0x1		// (9346CR)   page 3
-#define NIC_PAGE_STOP       0x2     // (PSTOP)    MSB, write-only
-#define NIC_BOUNDARY        0x3     // (BNRY)     MSB
-#define NIC_XMIT_START      0x4     // (TPSR)     MSB, write-only
-#define NIC_XMIT_STATUS     0x4     // (TSR)      read-only
-#define NIC_XMIT_COUNT_LSB  0x5     // (TBCR0)    write-only
-#define NIC_XMIT_COUNT_MSB  0x6     // (TBCR1)    write-only
-#define NIC_FIFO            0x6     // (FIFO)     read-only
-#define NIC_CONFIG3			0x6		// (CONFIG3)  
-#define NIC_INTR_STATUS     0x7     // (ISR)
-#define NIC_CURRENT         0x7     // (CURR)     page 1
-#define NIC_MC_ADDR         0x8     // (MAR0)     page 1
-#define NIC_CRDA_LSB        0x8     // (CRDA0)
-#define NIC_RMT_ADDR_LSB    0x8     // (RSAR0)
-#define NIC_CRDA_MSB        0x9     // (CRDA1)
-#define NIC_RMT_ADDR_MSB    0x9     // (RSAR1)
-#define NIC_RMT_COUNT_LSB   0xa     // (RBCR0)    write-only
-#define NIC_RMT_COUNT_MSB   0xb     // (RBCR1)    write-only
-#define NIC_RCV_CONFIG      0xc     // (RCR)      write-only
-#define NIC_RCV_STATUS      0xc     // (RSR)      read-only
-#define NIC_XMIT_CONFIG     0xd     // (TCR)      write-only
-#define NIC_FAE_ERR_CNTR    0xd     // (CNTR0)    read-only
-#define NIC_DATA_CONFIG     0xe     // (DCR)      write-only
-#define NIC_CRC_ERR_CNTR    0xe     // (CNTR1)    read-only
-#define NIC_INTR_MASK       0xf     // (IMR)      write-only
-#define NIC_MISSED_CNTR     0xf     // (CNTR2)    read-only
-#define NIC_RACK_NIC        0x10    // Byte to read or write
-#define NIC_RESET           0x1f    // (RESET)
-
-
-// Constants for the NIC_COMMAND register.
-//
-// Start/stop the card, start transmissions, and select
-// which page of registers was seen through the ports.
-
-#define CR_STOP         (UCHAR)0x01        // reset the card
-#define CR_START        (UCHAR)0x02        // start the card
-#define CR_XMIT         (UCHAR)0x04        // begin transmission
-#define CR_NO_DMA       (UCHAR)0x20        // stop remote DMA
-
-#define CR_PS0          (UCHAR)0x40        // low bit of page number
-#define CR_PS1          (UCHAR)0x80        // high bit of page number
-#define CR_PAGE0        (UCHAR)0x00        // select page 0
-#define CR_PAGE1        CR_PS0             // select page 1
-#define CR_PAGE2        CR_PS1             // select page 2
-#define CR_PAGE3        (CR_PS1+CR_PS0)    // select page 3
-
-#define CR_DMA_WRITE    (UCHAR)0x10        // Write
-#define CR_DMA_READ     (UCHAR)0x08        // Read
-#define CR_SEND         (UCHAR)0x18        // send
-
-
-// Constants for the NIC_XMIT_STATUS register.
-//
-// Indicate the result of a packet transmission.
-
-#define TSR_XMIT_OK     (UCHAR)0x01        // transmit with no errors
-#define TSR_COLLISION   (UCHAR)0x04        // collided at least once
-#define TSR_ABORTED     (UCHAR)0x08        // too many collisions
-#define TSR_NO_CARRIER  (UCHAR)0x10        // carrier lost
-#define TSR_NO_CDH      (UCHAR)0x40        // no collision detect heartbeat
-
-
-// Constants for the NIC_INTR_STATUS register.
-//
-// Indicate the cause of an interrupt.
-
-#define ISR_EMPTY       (UCHAR)0x00        // no bits set in ISR
-#define ISR_RCV         (UCHAR)0x01        // packet received with no errors
-#define ISR_XMIT        (UCHAR)0x02        // packet transmitted with no errors
-#define ISR_RCV_ERR     (UCHAR)0x04        // error on packet reception
-#define ISR_XMIT_ERR    (UCHAR)0x08        // error on packet transmission
-#define ISR_OVERFLOW    (UCHAR)0x10        // receive buffer overflow
-#define ISR_COUNTER     (UCHAR)0x20        // MSB set on tally counter
-#define ISR_DMA_DONE    (UCHAR)0x40        // RDC
-#define ISR_RESET       (UCHAR)0x80        // (not an interrupt) card is reset
-
-
-// Constants for the NIC_RCV_CONFIG register.
-//
-// Configure what type of packets are received.
-
-#define RCR_REJECT_ERR  (UCHAR)0x00        // reject error packets
-#define RCR_BROADCAST   (UCHAR)0x04        // receive broadcast packets
-#define RCR_MULTICAST   (UCHAR)0x08        // receive multicast packets
-#define RCR_ALL_PHYS    (UCHAR)0x10        // receive ALL directed packets
-#define RCR_MONITOR     (UCHAR)0x20        // don't collect packets
-
-
-// Constants for the NIC_RCV_STATUS register.
-//
-// Indicate the status of a received packet.
-//
-// These are also used to interpret the status byte in the
-// packet header of a received packet.
-
-#define RSR_PACKET_OK   (UCHAR)0x01        // packet received with no errors
-#define RSR_CRC_ERROR   (UCHAR)0x02        // packet received with CRC error
-#define RSR_MULTICAST   (UCHAR)0x20        // packet received was multicast
-#define RSR_DISABLED    (UCHAR)0x40        // received is disabled
-#define RSR_DEFERRING   (UCHAR)0x80        // receiver is deferring
-
-
-// Constants for the NIC_XMIT_CONFIG register.
-//
-// Configures how packets are transmitted.
-
-#define TCR_NO_LOOPBACK (UCHAR)0x00        // normal operation
-#define TCR_LOOPBACK    (UCHAR)0x02        // loopback (set when NIC is stopped)
-
-#define TCR_INHIBIT_CRC (UCHAR)0x01        // inhibit appending of CRC
-
-#define TCR_NIC_LBK     (UCHAR)0x02        // loopback through the NIC
-#define TCR_SNI_LBK     (UCHAR)0x04        // loopback through the SNI
-#define TCR_COAX_LBK    (UCHAR)0x06        // loopback to the coax
-
-
-// Constants for the NIC_DATA_CONFIG register.
-//
-// Set data transfer sizes.
-
-#define DCR_BYTE_WIDE   (UCHAR)0x00        // byte-wide DMA transfers
-#define DCR_WORD_WIDE   (UCHAR)0x01        // word-wide DMA transfers
-
-#define DCR_LOOPBACK    (UCHAR)0x00        // loopback mode (TCR must be set)
-#define DCR_NORMAL      (UCHAR)0x08        // normal operation
-
-#define DCR_FIFO_2_BYTE (UCHAR)0x00        // 2-byte FIFO threshhold
-#define DCR_FIFO_4_BYTE (UCHAR)0x20        // 4-byte FIFO threshhold
-#define DCR_FIFO_8_BYTE (UCHAR)0x40        // 8-byte FIFO threshhold
-#define DCR_FIFO_12_BYTE (UCHAR)0x60       // 12-byte FIFO threshhold
-#define DCR_AUTO_INIT   (UCHAR)0x10        // Auto-init to remove packets from ring
-
-
-// Constants for the NIC_INTR_MASK register.
-//
-// Configure which ISR settings actually cause interrupts.
-
-#define IMR_RCV         (UCHAR)0x01        // packet received with no errors
-#define IMR_XMIT        (UCHAR)0x02        // packet transmitted with no errors
-#define IMR_RCV_ERR     (UCHAR)0x04        // error on packet reception
-#define IMR_XMIT_ERR    (UCHAR)0x08        // error on packet transmission
-#define IMR_OVERFLOW    (UCHAR)0x10        // receive buffer overflow
-#define IMR_COUNTER     (UCHAR)0x20        // MSB set on tally counter
-
-#define XMIT_BUF_LEN			1536
-
-#define NE2000_NIC_XMIT_START	((UCHAR)(NE2000_RAM_BASE >> 8))
-
-// The start of the receive space.
-#define NE2000_NIC_PAGE_START	((UCHAR)((NE2000_RAM_BASE + XMIT_BUF_LEN) >> 8))
-
-// The end of the receive space.
-#define NE2000_NIC_PAGE_STOP	((UCHAR)((NE2000_RAM_BASE + NE2000_RAM_SIZE) >> 8))
-
-// Size of the ethernet header, usually 4 bytes, but bug times will be max 6
-#define NE2000_HEADER_SIZE          4
-
-// Field offset in the ethernet header
-#define HEADER_STATUS               0
-#define HEADER_NEXT_PACKET          1
-#define HEADER_PACKET_LEN_LOW	    2
-#define HEADER_PACKET_LEN_HIGH		3
-
diff -rupN '--exclude=.*' ../SDCC/include/ne2000.h ./include/ne2000.h
--- ../SDCC/include/ne2000.h	1970-01-01 02:00:00.000000000 +0200
+++ ./include/ne2000.h	2013-02-21 16:27:23.000000000 +0200
@@ -0,0 +1,186 @@
+// The header for an Novell 2000 driver.
+
+#define NE2000_RAM_BASE		0x4000
+#define NE2000_RAM_SIZE		0x2000
+
+//    Ne2000 Hardware definitions.
+
+// Offsets from Adapter->IoPAddr of the ports used to access
+// the 8390 NIC registers.
+//
+// The names in parenthesis are the abbreviations by which
+// the registers are referred to in the 8390 data sheet.
+//
+// Some of the offsets appear more than once
+// because they have have relevant page 0 and page 1 values,
+// or they are different registers when read than they are
+// when written. The notation MSB indicates that only the
+// MSB can be set for this register, the LSB is assumed 0.
+
+#define NIC_COMMAND         0x0     // (CR)
+#define NIC_PAGE_START      0x1     // (PSTART)   MSB, write-only
+#define NIC_PHYS_ADDR       0x1     // (PAR0)     page 1
+#define NIC_9346CR			0x1		// (9346CR)   page 3
+#define NIC_PAGE_STOP       0x2     // (PSTOP)    MSB, write-only
+#define NIC_BOUNDARY        0x3     // (BNRY)     MSB
+#define NIC_XMIT_START      0x4     // (TPSR)     MSB, write-only
+#define NIC_XMIT_STATUS     0x4     // (TSR)      read-only
+#define NIC_XMIT_COUNT_LSB  0x5     // (TBCR0)    write-only
+#define NIC_XMIT_COUNT_MSB  0x6     // (TBCR1)    write-only
+#define NIC_FIFO            0x6     // (FIFO)     read-only
+#define NIC_CONFIG3			0x6		// (CONFIG3)  
+#define NIC_INTR_STATUS     0x7     // (ISR)
+#define NIC_CURRENT         0x7     // (CURR)     page 1
+#define NIC_MC_ADDR         0x8     // (MAR0)     page 1
+#define NIC_CRDA_LSB        0x8     // (CRDA0)
+#define NIC_RMT_ADDR_LSB    0x8     // (RSAR0)
+#define NIC_CRDA_MSB        0x9     // (CRDA1)
+#define NIC_RMT_ADDR_MSB    0x9     // (RSAR1)
+#define NIC_RMT_COUNT_LSB   0xa     // (RBCR0)    write-only
+#define NIC_RMT_COUNT_MSB   0xb     // (RBCR1)    write-only
+#define NIC_RCV_CONFIG      0xc     // (RCR)      write-only
+#define NIC_RCV_STATUS      0xc     // (RSR)      read-only
+#define NIC_XMIT_CONFIG     0xd     // (TCR)      write-only
+#define NIC_FAE_ERR_CNTR    0xd     // (CNTR0)    read-only
+#define NIC_DATA_CONFIG     0xe     // (DCR)      write-only
+#define NIC_CRC_ERR_CNTR    0xe     // (CNTR1)    read-only
+#define NIC_INTR_MASK       0xf     // (IMR)      write-only
+#define NIC_MISSED_CNTR     0xf     // (CNTR2)    read-only
+#define NIC_RACK_NIC        0x10    // Byte to read or write
+#define NIC_RESET           0x1f    // (RESET)
+
+
+// Constants for the NIC_COMMAND register.
+//
+// Start/stop the card, start transmissions, and select
+// which page of registers was seen through the ports.
+
+#define CR_STOP         (UCHAR)0x01        // reset the card
+#define CR_START        (UCHAR)0x02        // start the card
+#define CR_XMIT         (UCHAR)0x04        // begin transmission
+#define CR_NO_DMA       (UCHAR)0x20        // stop remote DMA
+
+#define CR_PS0          (UCHAR)0x40        // low bit of page number
+#define CR_PS1          (UCHAR)0x80        // high bit of page number
+#define CR_PAGE0        (UCHAR)0x00        // select page 0
+#define CR_PAGE1        CR_PS0             // select page 1
+#define CR_PAGE2        CR_PS1             // select page 2
+#define CR_PAGE3        (CR_PS1+CR_PS0)    // select page 3
+
+#define CR_DMA_WRITE    (UCHAR)0x10        // Write
+#define CR_DMA_READ     (UCHAR)0x08        // Read
+#define CR_SEND         (UCHAR)0x18        // send
+
+
+// Constants for the NIC_XMIT_STATUS register.
+//
+// Indicate the result of a packet transmission.
+
+#define TSR_XMIT_OK     (UCHAR)0x01        // transmit with no errors
+#define TSR_COLLISION   (UCHAR)0x04        // collided at least once
+#define TSR_ABORTED     (UCHAR)0x08        // too many collisions
+#define TSR_NO_CARRIER  (UCHAR)0x10        // carrier lost
+#define TSR_NO_CDH      (UCHAR)0x40        // no collision detect heartbeat
+
+
+// Constants for the NIC_INTR_STATUS register.
+//
+// Indicate the cause of an interrupt.
+
+#define ISR_EMPTY       (UCHAR)0x00        // no bits set in ISR
+#define ISR_RCV         (UCHAR)0x01        // packet received with no errors
+#define ISR_XMIT        (UCHAR)0x02        // packet transmitted with no errors
+#define ISR_RCV_ERR     (UCHAR)0x04        // error on packet reception
+#define ISR_XMIT_ERR    (UCHAR)0x08        // error on packet transmission
+#define ISR_OVERFLOW    (UCHAR)0x10        // receive buffer overflow
+#define ISR_COUNTER     (UCHAR)0x20        // MSB set on tally counter
+#define ISR_DMA_DONE    (UCHAR)0x40        // RDC
+#define ISR_RESET       (UCHAR)0x80        // (not an interrupt) card is reset
+
+
+// Constants for the NIC_RCV_CONFIG register.
+//
+// Configure what type of packets are received.
+
+#define RCR_REJECT_ERR  (UCHAR)0x00        // reject error packets
+#define RCR_BROADCAST   (UCHAR)0x04        // receive broadcast packets
+#define RCR_MULTICAST   (UCHAR)0x08        // receive multicast packets
+#define RCR_ALL_PHYS    (UCHAR)0x10        // receive ALL directed packets
+#define RCR_MONITOR     (UCHAR)0x20        // don't collect packets
+
+
+// Constants for the NIC_RCV_STATUS register.
+//
+// Indicate the status of a received packet.
+//
+// These are also used to interpret the status byte in the
+// packet header of a received packet.
+
+#define RSR_PACKET_OK   (UCHAR)0x01        // packet received with no errors
+#define RSR_CRC_ERROR   (UCHAR)0x02        // packet received with CRC error
+#define RSR_MULTICAST   (UCHAR)0x20        // packet received was multicast
+#define RSR_DISABLED    (UCHAR)0x40        // received is disabled
+#define RSR_DEFERRING   (UCHAR)0x80        // receiver is deferring
+
+
+// Constants for the NIC_XMIT_CONFIG register.
+//
+// Configures how packets are transmitted.
+
+#define TCR_NO_LOOPBACK (UCHAR)0x00        // normal operation
+#define TCR_LOOPBACK    (UCHAR)0x02        // loopback (set when NIC is stopped)
+
+#define TCR_INHIBIT_CRC (UCHAR)0x01        // inhibit appending of CRC
+
+#define TCR_NIC_LBK     (UCHAR)0x02        // loopback through the NIC
+#define TCR_SNI_LBK     (UCHAR)0x04        // loopback through the SNI
+#define TCR_COAX_LBK    (UCHAR)0x06        // loopback to the coax
+
+
+// Constants for the NIC_DATA_CONFIG register.
+//
+// Set data transfer sizes.
+
+#define DCR_BYTE_WIDE   (UCHAR)0x00        // byte-wide DMA transfers
+#define DCR_WORD_WIDE   (UCHAR)0x01        // word-wide DMA transfers
+
+#define DCR_LOOPBACK    (UCHAR)0x00        // loopback mode (TCR must be set)
+#define DCR_NORMAL      (UCHAR)0x08        // normal operation
+
+#define DCR_FIFO_2_BYTE (UCHAR)0x00        // 2-byte FIFO threshhold
+#define DCR_FIFO_4_BYTE (UCHAR)0x20        // 4-byte FIFO threshhold
+#define DCR_FIFO_8_BYTE (UCHAR)0x40        // 8-byte FIFO threshhold
+#define DCR_FIFO_12_BYTE (UCHAR)0x60       // 12-byte FIFO threshhold
+#define DCR_AUTO_INIT   (UCHAR)0x10        // Auto-init to remove packets from ring
+
+
+// Constants for the NIC_INTR_MASK register.
+//
+// Configure which ISR settings actually cause interrupts.
+
+#define IMR_RCV         (UCHAR)0x01        // packet received with no errors
+#define IMR_XMIT        (UCHAR)0x02        // packet transmitted with no errors
+#define IMR_RCV_ERR     (UCHAR)0x04        // error on packet reception
+#define IMR_XMIT_ERR    (UCHAR)0x08        // error on packet transmission
+#define IMR_OVERFLOW    (UCHAR)0x10        // receive buffer overflow
+#define IMR_COUNTER     (UCHAR)0x20        // MSB set on tally counter
+
+#define XMIT_BUF_LEN			1536
+
+#define NE2000_NIC_XMIT_START	((UCHAR)(NE2000_RAM_BASE >> 8))
+
+// The start of the receive space.
+#define NE2000_NIC_PAGE_START	((UCHAR)((NE2000_RAM_BASE + XMIT_BUF_LEN) >> 8))
+
+// The end of the receive space.
+#define NE2000_NIC_PAGE_STOP	((UCHAR)((NE2000_RAM_BASE + NE2000_RAM_SIZE) >> 8))
+
+// Size of the ethernet header, usually 4 bytes, but bug times will be max 6
+#define NE2000_HEADER_SIZE          4
+
+// Field offset in the ethernet header
+#define HEADER_STATUS               0
+#define HEADER_NEXT_PACKET          1
+#define HEADER_PACKET_LEN_LOW	    2
+#define HEADER_PACKET_LEN_HIGH		3
+
diff -rupN '--exclude=.*' ../SDCC/include/version.h ./include/version.h
--- ../SDCC/include/version.h	2013-02-16 20:28:52.000000000 +0200
+++ ./include/version.h	2013-02-22 13:29:50.000000000 +0200
@@ -8,10 +8,10 @@
 //#define VER_GP1266			// Digitmat extended GPIO key version with PoE support
 //#define VER_BT2008			// BitCall extended GPIO key version with PoE support, 2x16 LCD
 //#define VER_AR168J			// Hardware version 6.00
-#define VER_AR168K			// Hardware version 7.00, 2x16 LCD, 5x6 key
+//#define VER_AR168K			// Hardware version 7.00, 2x16 LCD, 5x6 key
 //#define VER_AR168KM			// 2x16 LCD, 5x6 key, Micrel KSZ8842, SST39VF1681
 //#define VER_GP2266			// Digitmat AR168K with dot matrix display 
-//#define VER_AR168M			// AR1688 VoIP Module
+#define VER_AR168M			// AR1688 VoIP Module
 //#define VER_AR168MK			// AR1688 VoIP Module with SST39VF1681, Micrel KSZ8842
 //#define VER_AR168MS			// AR1688 VoIP Module with SST39VF1681
 //#define VER_AR168MT			// AR1688 VoIP Module with Micrel KSZ8842
@@ -36,7 +36,7 @@
 //#define RES_BR	// Brazil (Portuguese)
 //#define RES_CA	// Canada
 //#define RES_CL	// Chile
-#define RES_CN		// China (Mandarin) ***
+//#define RES_CN		// China (Mandarin) ***
 //#define RES_CZ	// Czech Republic
 //#define RES_DE	// Germany
 //#define RES_EG	// Egypt
@@ -63,7 +63,7 @@
 //#define RES_SK	// Slovakia
 //#define RES_TR	// Turkey
 //#define RES_TW	// China Taiwan, use Traditional Chinese.
-//#define RES_US	// United States (English) ***
+#define RES_US	// United States (English) ***
 //#define RES_VN	// Vietnam
 //#define RES_ZA	// South Africa
 
@@ -413,3 +413,7 @@
 #define ENG_BUILD_HIGH		0
 #define ENG_BUILD_MID		0
 #define ENG_BUILD_LOW		2
+
+
+
+
diff -rupN '--exclude=.*' ../SDCC/src/makefile ./src/makefile
--- ../SDCC/src/makefile	2013-02-16 20:13:12.000000000 +0200
+++ ./src/makefile	2013-02-21 16:27:23.000000000 +0200
@@ -1,3 +1,4 @@
+ifdef SystemRoot
 PATH_SDCC=C:\SDCC
 
 RM = del
@@ -10,7 +11,21 @@ CZ80_OPT=-mz80 -c --std-c99 --codeseg CO
 AZ80_OPT=-plosgff
 
 VERSION_FILE = $(PATH_SDCC)\include\version.h
+LINKMAIN = linkmain
+else
+RM = rm
+AZ80 = sdasz80
+CZ80 = sdcc
+LZ80 = sdld
+HEX2BIN = ../bin/hex2bin
+INC = -I../include
 
+CZ80_OPT=-mz80 $(INC) -c --std-c99 --codeseg CODE
+AZ80_OPT=-plosgff
+
+VERSION_FILE = ../include/version.h
+LINKMAIN = linkmain_unix
+endif
 %.rel:	%.c
 	$(CZ80) $*.c $(CZ80_OPT)
 
@@ -182,10 +197,10 @@ sipr.rel: sipr.c $(VERSION_FILE)
 	$(CZ80) sipr.c $(CZ80_OPT)7
 
 main.ihx: $(SRAM_OBJS) $(FONT_OBJS) $(FLASH0_OBJS) $(FLASH1_OBJS) $(FLASH2_OBJS) $(FLASH3_OBJS) $(FLASH4_OBJS) $(FLASH5_OBJS) $(FLASH6_OBJS) $(FLASH7_OBJS)
-	$(LZ80) -n -f linkmain
+	$(LZ80) -n -f $(LINKMAIN)
 
 main.bin: main.ihx
 	$(HEX2BIN) main.bin -f0 -a0 -x6 -b32768 -e57344 main.ihx
 	
 clean:
-	$(RM) *.bak *.lst *.rel *.ihx *.sym *.asm *.map *.lnk *.~*
+	$(RM) *.lst *.rel *.ihx *.sym *.asm *.map *.bin *.noi linkmain_unix.lnk
diff -rupN '--exclude=.*' ../SDCC/src/menu.c ./src/menu.c
--- ../SDCC/src/menu.c	2012-08-12 23:16:24.000000000 +0300
+++ ./src/menu.c	2013-02-21 16:27:23.000000000 +0200
@@ -1301,15 +1301,15 @@ void CallLogAddToPb()
  *******************************************************************
 */
 #if defined RES_RO
-#include "res\\ro\\inputmap.h"
+#include "res/ro/inputmap.h"
 #elif defined RES_RU
-#include "res\\ru\\inputmap.h"
+#include "res/ru/inputmap.h"
 #elif defined RES_TR && defined OEM_IP20
-#include "res\\ip20\\tr\\inputmap.h"
+#include "res/ip20/tr/inputmap.h"
 #elif defined RES_VN
-#include "res\\vn\\inputmap.h"
+#include "res/vn/inputmap.h"
 #else // RES_US and others
-#include "res\\us\\inputmap.h"
+#include "res/us/inputmap.h"
 #endif
 
 void InputNext()
diff -rupN '--exclude=.*' ../SDCC/src/mk ./src/mk
--- ../SDCC/src/mk	1970-01-01 02:00:00.000000000 +0200
+++ ./src/mk	2013-02-21 16:27:23.000000000 +0200
@@ -0,0 +1,5 @@
+#!/bin/bash
+../bin/reversion ../include/version.h $1 $2 $3 $4
+../bin/relink linkmain.lnk $1 $2 $3 $4
+make
+../bin/namebin main.bin $1 $2 $3 $4
diff -rupN '--exclude=.*' ../SDCC/src/time.c ./src/time.c
--- ../SDCC/src/time.c	2012-06-29 12:33:28.000000000 +0300
+++ ./src/time.c	2013-02-21 16:27:23.000000000 +0200
@@ -80,30 +80,30 @@
 
 
 #ifdef RES_BR // Brasil
-#include "res\\br\\time.h"
+#include "res/br/time.h"
 #elif defined RES_CN && defined DISPLAY_DOT
 #elif defined RES_ES // Español
-#include "res\\es\\time.h"
+#include "res/es/time.h"
 #elif defined RES_FR // French
-#include "res\\fr\\time.h"
+#include "res/fr/time.h"
 #elif defined RES_IT // Italy
-#include "res\\it\\time.h"
+#include "res/it/time.h"
 #elif defined RES_KR && defined DISPLAY_DOT // Korean
-#include "res\\kr\\time.h"
+#include "res/kr/time.h"
 #elif defined RES_RO // Romanian
-#include "res\\ro\\time.h"
+#include "res/ro/time.h"
 #elif defined RES_RU && defined DISPLAY_DOT // Russian
-#include "res\\ru\\time.h"
+#include "res/ru/time.h"
 #elif defined RES_TR // Turkish
 #ifdef OEM_IP20
-#include "res\\ip20\\tr\\time.h"
+#include "res/ip20/tr/time.h"
 #else
-#include "res\\tr\\time.h"
+#include "res/tr/time.h"
 #endif
 #elif defined RES_VN // Vietnam
-#include "res\\vn\\time.h"
+#include "res/vn/time.h"
 #else // Default = REG_US
-#include "res\\us\\time.h"
+#include "res/us/time.h"
 #endif
 
 const UCHAR _cMonthDay[MONTHS_PER_YEAR] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
diff -rupN '--exclude=.*' ../SDCC/src/ui_str.c ./src/ui_str.c
--- ../SDCC/src/ui_str.c	2012-06-29 12:33:26.000000000 +0300
+++ ./src/ui_str.c	2013-02-21 16:27:23.000000000 +0200
@@ -45,39 +45,39 @@
 
 #ifdef OEM_BT2008
 #if defined RES_CN && defined DISPLAY_DOT
-#include "res\\bt2008\\cn\\menu.h"
+#include "res/bt2008/cn/menu.h"
 #else // Default = REG_US
-#include "res\\bt2008\\us\\menu.h"
+#include "res/bt2008/us/menu.h"
 #endif
 #elif defined OEM_INNOMEDIA
-#include "res\\innomedia\\us\\menu.h"
+#include "res/innomedia/us/menu.h"
 #else
 #ifdef RES_BR // Brasil
-#include "res\\br\\menu.h"
+#include "res/br/menu.h"
 #elif defined RES_CN && defined DISPLAY_DOT
-#include "res\\cn\\menu.h"
+#include "res/cn/menu.h"
 #elif defined RES_ES // Español
-#include "res\\es\\menu.h"
+#include "res/es/menu.h"
 #elif defined RES_FR // French
-#include "res\\fr\\menu.h"
+#include "res/fr/menu.h"
 #elif defined RES_IT // Italy
-#include "res\\it\\menu.h"
+#include "res/it/menu.h"
 #elif defined RES_KR && defined DISPLAY_DOT // Korean
-#include "res\\kr\\menu.h"
+#include "res/kr/menu.h"
 #elif defined RES_RO // Romanian
-#include "res\\ro\\menu.h"
+#include "res/ro/menu.h"
 #elif defined RES_RU && defined DISPLAY_DOT // Russian
-#include "res\\ru\\menu.h"
+#include "res/ru/menu.h"
 #elif defined RES_TR // Turkish
 #ifdef OEM_IP20
-#include "res\\ip20\\tr\\menu.h"
+#include "res/ip20/tr/menu.h"
 #else
-#include "res\\tr\\menu.h"
+#include "res/tr/menu.h"
 #endif
 #elif defined RES_VN // Vietnam
-#include "res\\vn\\menu.h"
+#include "res/vn/menu.h"
 #else // Default = REG_US
-#include "res\\us\\menu.h"
+#include "res/us/menu.h"
 #endif
 #endif
 
@@ -85,39 +85,39 @@ const UCHAR _cIDS[][DISPLAY_MAX_CHAR+1]
 {'V', 'e', 'r', ':', ' ', SOFT_VER_HIGH+'0', '.', SOFT_VER_LOW+'0', SOFT_VER_BUILD+'0', '.', ENG_BUILD_HIGH+'0', ENG_BUILD_MID+'0', ENG_BUILD_LOW+'0', 0},
 #ifdef OEM_BT2008
 #if defined RES_CN && defined DISPLAY_DOT
-#include "res\\bt2008\\cn\\str.h"
+#include "res/bt2008/cn/str.h"
 #else // Default = REG_US
-#include "res\\bt2008\\us\\str.h"
+#include "res/bt2008/us/str.h"
 #endif
 #elif defined OEM_INNOMEDIA
-#include "res\\innomedia\\us\\str.h"
+#include "res/innomedia/us/str.h"
 #else
 #ifdef RES_BR // Brasil
-#include "res\\br\\str.h"
+#include "res/br/str.h"
 #elif defined RES_CN && defined DISPLAY_DOT
-#include "res\\cn\\str.h"
+#include "res/cn/str.h"
 #elif defined RES_ES // Español
-#include "res\\es\\str.h"
+#include "res/es/str.h"
 #elif defined RES_FR // French
-#include "res\\fr\\str.h"
+#include "res/fr/str.h"
 #elif defined RES_IT // Italy
-#include "res\\it\\str.h"
+#include "res/it/str.h"
 #elif defined RES_KR && defined DISPLAY_DOT // Korean
-#include "res\\kr\\str.h"
+#include "res/kr/str.h"
 #elif defined RES_RO // Romanian
-#include "res\\ro\\str.h"
+#include "res/ro/str.h"
 #elif defined RES_RU && defined DISPLAY_DOT // Russian
-#include "res\\ru\\str.h"
+#include "res/ru/str.h"
 #elif defined RES_TR // Turkish
 #ifdef OEM_IP20
-#include "res\\ip20\\tr\\str.h"
+#include "res/ip20/tr/str.h"
 #else
-#include "res\\tr\\str.h"
+#include "res/tr/str.h"
 #endif
 #elif defined RES_VN // Vietnam
-#include "res\\vn\\str.h"
+#include "res/vn/str.h"
 #else // Default = REG_US
-#include "res\\us\\str.h"
+#include "res/us/str.h"
 #endif
 #endif
 };
diff -rupN '--exclude=.*' ../SDCC/tool/Makefile ./tool/Makefile
--- ../SDCC/tool/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/Makefile	2013-02-22 12:00:04.000000000 +0200
@@ -0,0 +1,16 @@
+DIRS = reversion relink hex2bin namebin
+
+all clean:
+	@for dir in $(DIRS); do \
+		if $(MAKE) -C $$dir $@; then \
+			true; \
+		else \
+			exit 1; \
+		fi; \
+	done
+
+install:
+	@for dir in $(DIRS); do \
+		echo "*** Copy $$dir to bin"; \
+		cp -RLf $$dir/$$dir ../bin/; \
+	done
diff -rupN '--exclude=.*' ../SDCC/tool/common/cmd_default.cpp ./tool/common/cmd_default.cpp
--- ../SDCC/tool/common/cmd_default.cpp	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/common/cmd_default.cpp	2013-02-22 10:06:15.000000000 +0200
@@ -1,4 +1,6 @@
-#include "stdafx.h"
+
+#include <StdAfx.h>
+
 
 void CommandLineDefault1(int argc, TCHAR* argv[], CString & strVer, CString & strCall, CString & strRes, CString & strOem)
 {
diff -rupN '--exclude=.*' ../SDCC/tool/common/common.cpp ./tool/common/common.cpp
--- ../SDCC/tool/common/common.cpp	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/common/common.cpp	2013-02-22 14:09:00.000000000 +0200
@@ -1,18 +1,29 @@
+
 #include <stdafx.h>
 
 BOOL PromptOverwriteFile(CString strFileName, BOOL bForce)
 {
 	char pAnswer[80];
 	CString strAnswer;
+#ifndef __GNUC__
 	CFileStatus status;
 
 	if (bForce)									return TRUE;
 	if (!CFile::GetStatus(strFileName, status))	return TRUE;
+#else
+	struct stat status;
 
+	if (bForce)									return TRUE;
+   if (stat(strFileName.c_str(),&status)) return TRUE;
+#endif
 	while (1)
 	{
 		printf("The destination file exists. Overwrite the file?(Yes/No):");
+#ifndef __GNUC__
 		gets(pAnswer);
+#else
+      fgets( pAnswer, sizeof(pAnswer), stdin );
+#endif
 		strAnswer = pAnswer;
 		if (!strAnswer.CompareNoCase(_T("no")) || !strAnswer.CompareNoCase(_T("n")))
 		{
diff -rupN '--exclude=.*' ../SDCC/tool/common/mfc2std.h ./tool/common/mfc2std.h
--- ../SDCC/tool/common/mfc2std.h	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/common/mfc2std.h	2013-02-22 09:17:25.000000000 +0200
@@ -0,0 +1,154 @@
+/* 
+   mcf2std.h : MCF to C++ convertion helper for Palmmicro tools compilation on *UNIX
+   by Alex Vangelov
+*/
+
+#ifdef __GNUC__
+#define TRUE      true
+#define FALSE     false
+#define BOOL      bool
+#define _tsystem  system
+#define _wtoi(x)  atoi(x.c_str())
+#ifdef _UNICODE
+#define T(x) L##x
+#else
+#define _T(x) x
+#endif
+#define GetCurrentDirectory(l,x) getcwd(x, l) 
+#define modeRead  in
+#define typeText  in
+#define modeCreate   out
+#define modeWrite out
+#define POSITION        std::list<CString>::iterator
+#define WideCharToMultiByte
+#define CFileException  std::exception
+
+#include <string>
+#include <fstream>
+#include <cstring>
+#include <list>
+#include <stdarg.h>
+#include <algorithm>
+#include <unistd.h>
+
+typedef unsigned char BYTE;
+typedef bool BOOLEAN;
+typedef char TCHAR;
+typedef const TCHAR *LPCTSTR;
+typedef unsigned int UINT;
+
+class CString : public std::string 
+{
+public:
+   CString() : std::string() { }
+	CString( const std::string& s ) : std::string( s ) { }
+	CString( const std::string& s, std::size_t n ) : std::string( s,n ) { }
+	CString( const char *s, std::size_t n ) : std::string( s,n ) { }
+	CString( const char *s ) : std::string( s ) { }
+	CString( std::size_t n, char c ) : std::string ( n,c ) { }
+
+   CString Left(size_t n) {
+      return substr(0,n < length() ? n : length() ).c_str();
+   }
+   CString Right(size_t n) {
+      size_t l = length() < n ? length() : n;
+      return substr(length()-l,length()).c_str();
+   }
+   void MakeLower() {
+      std::transform(begin(), end(), begin(), ::tolower);
+   }
+   void MakeUpper() {
+      std::transform(begin(), end(), begin(), ::toupper);    
+   }
+   size_t GetLength() { 
+      size_t l = length();
+      return l; 
+   }
+   char GetAt(size_t pos) { 
+      return at(pos); 
+   }
+   void TrimLeft() {
+      erase(begin(), std::find_if(begin(), end(), std::not1(std::ptr_fun<int, int>(std::isspace))));
+   }
+   bool CompareNoCase(CString s) {
+      MakeLower();
+      s.MakeLower();
+      return compare(s);
+   }
+   CString Mid(size_t pos = 0, size_t len = npos) {
+      return substr(pos,len);
+   }
+   void FindAndReplace(CString search, CString replace) {
+     size_t pos = 0;
+      while((pos = find(search, pos)) != std::string::npos) {
+         this->replace(pos, search.length(), replace);
+         pos += replace.length();
+      }
+   }
+
+   operator const char*(){
+		return c_str();
+	}
+
+};
+
+class CStdioFile : public std::fstream 
+{
+public:
+   bool Open(CString filename, ios_base::openmode mode) {
+      open(filename.c_str(), mode);
+      return is_open();
+   }
+   void Close() { 
+      close(); 
+   }
+   void WriteString(CString str) {
+      write(str, str.length());   
+   }
+};
+
+class CStringList : public std::list<CString> 
+{
+public:
+   void AddTail(CString str) {
+      push_back(str);
+   }
+   CString GetNext(POSITION pos) {
+      CString next = *(pos++);
+      return next;
+   }
+   void SetAt(POSITION pos, CString str) {
+      *pos = str;   
+   }
+};
+
+class CFile : public std::fstream 
+{
+public:
+   bool Open(CString filename, ios_base::openmode mode) {
+      open(filename.c_str(), mode | std::ios::binary);
+      return is_open();
+   }
+   int GetLength() {
+      int length;
+      seekg (0, std::ios::end);
+      length = tellg();
+      seekg (0, std::ios::beg);   
+      return length;
+   }
+   void Read(char* s, size_t n) {
+      read(s,n);
+   }
+   void Write(char* s, size_t n) {
+      write(s,n);
+   }
+   void Close() { 
+      close(); 
+   }
+   static void Remove(CString f) {
+       remove(f); 
+   }
+};
+
+
+#endif
diff -rupN '--exclude=.*' ../SDCC/tool/hex2bin/HexFile.cpp ./tool/hex2bin/HexFile.cpp
--- ../SDCC/tool/hex2bin/HexFile.cpp	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/hex2bin/HexFile.cpp	2013-02-22 10:40:11.000000000 +0200
@@ -1,7 +1,8 @@
 // HexFile.cpp : implementation file
 //
 
-#include "stdafx.h"
+#include "StdAfx.h"
+
 #include "HexFile.h"
 
 #ifdef _DEBUG
@@ -13,25 +14,39 @@ static char THIS_FILE[] = __FILE__;
 /////////////////////////////////////////////////////////////////////////////
 // CHexFile
 
+#ifndef __GNUC__
 IMPLEMENT_DYNAMIC(CHexFile, CStdioFile)
-
+#endif
 CHexFile::CHexFile() : CStdioFile()
 {
 }
 
+#ifndef __GNUC__
 BOOL CHexFile::Open(LPCTSTR lpszFileName, UINT nOpenFlags, CFileException* pError)
 {
 	return CStdioFile::Open(lpszFileName, nOpenFlags, pError);
 }
+#endif
 
 BOOL CHexFile::ReadLine()
 {
+#ifndef __GNUC__
 	if (!ReadString(m_strLine))		return FALSE;
+#else
+   char buff[512];
+   if (!getline(buff,511)) return FALSE;
+   if (bad()) return FALSE;
+   m_strLine = buff;
+#endif
 
 	int i, iVal, iCheckSum = 0;
 //	const char * p = LPCTSTR(m_strLine);
 	char p[1024];
+#ifndef __GNUC__
 	WideCharToMultiByte(CP_ACP, 0, m_strLine, -1, p, 1024, NULL, NULL); 
+#else
+   memcpy(p, m_strLine.c_str(), m_strLine.length());
+#endif
 	
 	if (p[0] != ':' || p[7] != '0' || p[8] != '0')	return FALSE;
 
diff -rupN '--exclude=.*' ../SDCC/tool/hex2bin/HexFile.h ./tool/hex2bin/HexFile.h
--- ../SDCC/tool/hex2bin/HexFile.h	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/hex2bin/HexFile.h	2013-02-21 16:27:23.000000000 +0200
@@ -16,11 +16,14 @@
 
 class CHexFile : public CStdioFile
 {
+#ifndef __GNUC__
 	DECLARE_DYNAMIC(CHexFile)
+#endif
 public:
 	CHexFile();
-
+#ifndef __GNUC__
 	virtual BOOL Open(LPCTSTR lpszFileName, UINT nOpenFlags = CFile::modeRead|CFile::typeText, CFileException* pError = NULL);
+#endif
 	BOOL ReadLine();
 
 	inline int GetNumber()	{	return m_iNumber;	}
diff -rupN '--exclude=.*' ../SDCC/tool/hex2bin/Makefile ./tool/hex2bin/Makefile
--- ../SDCC/tool/hex2bin/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/hex2bin/Makefile	2013-02-22 11:07:00.000000000 +0200
@@ -0,0 +1,14 @@
+CC=g++
+CFLAGS=-I.
+OBJ = hex2bin.o HexFile.o 
+
+all: hex2bin
+
+%.o: %.cpp
+	$(CC) -c -o $@ $< $(CFLAGS)
+
+hex2bin: $(OBJ)
+	g++ -o $@ $^ $(CFLAGS)
+
+clean:
+	rm -f *.o *~ hex2bin
diff -rupN '--exclude=.*' ../SDCC/tool/hex2bin/StdAfx.h ./tool/hex2bin/StdAfx.h
--- ../SDCC/tool/hex2bin/StdAfx.h	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/hex2bin/StdAfx.h	2013-02-22 10:40:02.000000000 +0200
@@ -6,6 +6,8 @@
 #if !defined(AFX_STDAFX_H__C09060B0_3E8D_40A1_8AEB_A975643894FA__INCLUDED_)
 #define AFX_STDAFX_H__C09060B0_3E8D_40A1_8AEB_A975643894FA__INCLUDED_
 
+#ifndef __GNUC__
+
 #if _MSC_VER > 1000
 #pragma once
 #endif // _MSC_VER > 1000
@@ -30,7 +32,15 @@
 
 // TODO: reference additional headers your program requires here
 
+#include "..\\..\\include\\ar168.h"
 //{{AFX_INSERT_LOCATION}}
 // Microsoft Visual C++ will insert additional declarations immediately before the previous line.
 
+#else // __GNUC__
+#include "../common/mfc2std.h"
+#include "../common/common.h"
+#include "../../include/version.h"
+#include "../../include/ar168.h"
+#endif
+
 #endif // !defined(AFX_STDAFX_H__C09060B0_3E8D_40A1_8AEB_A975643894FA__INCLUDED_)
diff -rupN '--exclude=.*' ../SDCC/tool/hex2bin/hex2bin.cpp ./tool/hex2bin/hex2bin.cpp
--- ../SDCC/tool/hex2bin/hex2bin.cpp	2012-06-29 12:33:30.000000000 +0300
+++ ./tool/hex2bin/hex2bin.cpp	2013-02-22 10:40:16.000000000 +0200
@@ -1,11 +1,9 @@
 // hex2bin.cpp : Defines the entry point for the console application.
 //
 
-#include "stdafx.h"
+#include "StdAfx.h"
 
-#include "Hexfile.h"
-
-#include "../../include/ar168.h"
+#include "HexFile.h"
 
 void MainProcess(int argc, TCHAR* argv[]);
 
@@ -17,7 +15,7 @@ static char THIS_FILE[] = __FILE__;
 
 /////////////////////////////////////////////////////////////////////////////
 // The one and only application object
-
+#ifndef __GNUC__
 CWinApp theApp;
 
 using namespace std;
@@ -35,6 +33,12 @@ int _tmain(int argc, TCHAR* argv[], TCHA
 	}
 	else
 	{
+#else
+int main(int argc, char *argv[])
+{
+   int nRetCode = 0;
+   {
+#endif
 		// TODO: code your application's behavior here.
 		if (argc < 3)
 		{
@@ -67,7 +71,11 @@ void MainProcess(int argc, TCHAR* argv[]
 	{
 		if (argv[i][0] == _T('-'))
 		{
+#ifndef __GNUC__
 			WideCharToMultiByte(CP_ACP, 0, argv[i], -1, pArgv, 128, NULL, NULL); 
+#else
+         memcpy(pArgv, argv[i], 128);
+#endif
 			switch (argv[i][1])
 			{
 			case _T('f'):
@@ -102,7 +110,12 @@ void MainProcess(int argc, TCHAR* argv[]
 			memset(p, iFillChar, HEXFILE_SIZE);
 			bCommonBank = FALSE;
 			bError = FALSE;
+#ifndef __GNUC__
 			if (hex.Open(argv[i]))
+#else
+         CString tmp = argv[i];
+         if (hex.Open(tmp, std::fstream::in))
+#endif
 			{
 				while (hex.ReadLine())
 				{
diff -rupN '--exclude=.*' ../SDCC/tool/namebin/Makefile ./tool/namebin/Makefile
--- ../SDCC/tool/namebin/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/namebin/Makefile	2013-02-22 11:07:13.000000000 +0200
@@ -0,0 +1,14 @@
+CC=g++
+CFLAGS=-I.
+OBJ = namebin.o ../common/cmd_default.o
+
+all: namebin
+
+%.o: %.cpp
+	$(CC) -c -o $@ $< $(CFLAGS)
+
+namebin: $(OBJ)
+	g++ -o $@ $^ $(CFLAGS)
+
+clean:
+	rm -f *.o *~ namebin
diff -rupN '--exclude=.*' ../SDCC/tool/namebin/StdAfx.h ./tool/namebin/StdAfx.h
--- ../SDCC/tool/namebin/StdAfx.h	2012-06-29 12:33:32.000000000 +0300
+++ ./tool/namebin/StdAfx.h	2013-02-22 14:04:10.000000000 +0200
@@ -6,6 +6,8 @@
 #if !defined(AFX_STDAFX_H__8692AECA_5BC3_4136_9B64_74A4DC55AE75__INCLUDED_)
 #define AFX_STDAFX_H__8692AECA_5BC3_4136_9B64_74A4DC55AE75__INCLUDED_
 
+#ifndef __GNUC__
+
 #if _MSC_VER > 1000
 #pragma once
 #endif // _MSC_VER > 1000
@@ -30,7 +32,18 @@
 
 // TODO: reference additional headers your program requires here
 
+#include "..\\common\\common.h"
+#include "..\\..\\include\\ar168.h"
+
 //{{AFX_INSERT_LOCATION}}
 // Microsoft Visual C++ will insert additional declarations immediately before the previous line.
 
+#else // __GNUC__
+#include <sys/stat.h>
+#include "../common/mfc2std.h"
+#include "../common/common.h"
+#include "../../include/version.h"
+#include "../../include/ar168.h"
+#endif
+
 #endif // !defined(AFX_STDAFX_H__8692AECA_5BC3_4136_9B64_74A4DC55AE75__INCLUDED_)
diff -rupN '--exclude=.*' ../SDCC/tool/namebin/namebin.cpp ./tool/namebin/namebin.cpp
--- ../SDCC/tool/namebin/namebin.cpp	2012-12-15 21:50:32.000000000 +0200
+++ ./tool/namebin/namebin.cpp	2013-02-22 10:42:29.000000000 +0200
@@ -1,10 +1,7 @@
 // namebin.cpp : Defines the entry point for the console application.
 //
 
-#include "stdafx.h"
-#include "..\\..\\include\\ar168.h"
-
-#include "..\\common\\common.h"
+#include "StdAfx.h"
 
 #ifdef _DEBUG
 #define new DEBUG_NEW
@@ -23,8 +20,9 @@ void _SetFile(char * p, CString strVer,
 
 /////////////////////////////////////////////////////////////////////////////
 // The one and only application object
-
+#ifndef __GNUC__
 CWinApp theApp;
+#endif
 
 using namespace std;
 
@@ -37,7 +35,7 @@ void PrintHelp()
 	printf("%s\n", "[country]\tPhone's language");
 	printf("%s\n", "[oem]\t\tOEM. use 'size' to report code size, use 'zip' to zip output .bin file");
 }
-
+#ifndef __GNUC__
 int _tmain(int argc, TCHAR* argv[], TCHAR* envp[])
 {
 	int nRetCode = 0;
@@ -51,7 +49,13 @@ int _tmain(int argc, TCHAR* argv[], TCHA
 	}
 	else
 	{
-		printf("Palmmicro AR1688 namebin.exe utility 0.59.028\n");
+#else
+int main(int argc, char *argv[])
+{
+   int nRetCode = 0;
+   {
+#endif
+		printf("Palmmicro AR1688 namebin utility 0.59.028\n");
 		if (argc < 2)
 		{
 			PrintHelp();
@@ -93,16 +97,28 @@ CString _GetVersion(CString strCurDir)
 	CString strIn;
 	CString str;
 	BOOL bStart;
-
+#ifndef __GNUC__
 	if (!in.Open(strCurDir.Left(strCurDir.GetLength() - 4) + _T("include\\version.h"), CFile::modeRead|CFile::typeText))
+#else
+	if (!in.Open(strCurDir.Left(strCurDir.GetLength() - 4) + _T("include/version.h"), CFile::modeRead|CFile::typeText))
+#endif
 	{
 		goto End;
-	}
+	} 
 
 	strVersion = "";
 	bStart = FALSE;
+#ifndef __GNUC__
 	while (in.ReadString(strIn))
 	{
+#else
+   while (!in.eof())
+   {
+      CString strIn;
+      std::getline(in, strIn);
+      
+      if (*strIn.rbegin()== '\r') strIn.erase(strIn.length() - 1);
+#endif
 		if (strIn.Left(10) == _T("/*########"))
 		{
 			bStart = TRUE;
@@ -152,7 +168,6 @@ CString _GetVersion(CString strCurDir)
 			}
 		}
 	}
-
 	in.Close();
 
 End:
@@ -165,7 +180,11 @@ void _SetFlag(char * p, CString strVal,
 	int iLen;
 
 	iLen = strVal.GetLength();
+#ifndef __GNUC__
 	WideCharToMultiByte(CP_ACP, 0, strVal, -1, p, iMax, NULL, NULL); 
+#else
+   memcpy(p, strVal.c_str(), iLen);
+#endif
 //	memcpy(p, strVal, iLen);
 /*	for (i = 0; i < iLen; i ++)
 	{
@@ -218,7 +237,11 @@ BOOL _WriteHttpPage(char * p, CStringLis
 
 	iIndex = 0;
 	iAddr = 64;
+#ifndef __GNUC__
 	for (pos = listSrcName.GetHeadPosition(); pos != NULL;)
+#else
+   for (pos=listSrcName.begin(); pos != listSrcName.end(); ++pos)
+#endif
 	{
 		strSrcName = listSrcName.GetNext(pos);
 		if (!in.Open(strSrcName, CFile::modeRead))
@@ -261,16 +284,28 @@ BOOL _MergeFile(CString strCurDir, CStri
 	char * pCur;
 	CString strTempFile;
 	CString strTempVer;
+#ifndef __GNUC__
 	CFileStatus status;
+#else
+   struct stat status;
+#endif
 	
 	if (!in.Open(strSrcName, CFile::modeRead))
 	{
+#ifndef __GNUC__
 		wprintf(_T("\nUnable to open %s"), strSrcName);
+#else
+      printf(_T("\nUnable to open %s"), strSrcName.c_str());
+#endif
 		return FALSE;
 	}
 	if (!out.Open(strDstName, CFile::modeCreate|CFile::modeWrite))
 	{
+#ifndef __GNUC__
 		wprintf(_T("\nUnable to create %s"), strDstName);
+#else
+      printf(_T("\nUnable to create %s"), strDstName.c_str());
+#endif
 		in.Close();
 		return FALSE;
 	}
@@ -291,7 +326,11 @@ BOOL _MergeFile(CString strCurDir, CStri
 	if (iLength > FILE_FULL_PAGE_SIZE)
 	{	
 		// add http data for main upgrade binary file
+#ifndef __GNUC__
 		strResDir = strCurDir + _T("res\\");
+#else
+      strResDir = strCurDir + _T("res/");
+#endif
 		strTempVer = strVer;
 		strTempVer.MakeLower();
 		if (strOem == _T("bt2008") || strTempVer == _T("bt2008") || strTempVer == _T("bt2008n"))
@@ -300,22 +339,44 @@ BOOL _MergeFile(CString strCurDir, CStri
 		}
 		else if (strOem == _T("innomedia"))
 		{
+#ifndef __GNUC__
 			strResDir += _T("innomedia\\");
+#else
+         strResDir += _T("innomedia/");
+#endif
 		}
 		else if (strOem == _T("ip20"))
 		{
+#ifndef __GNUC__
 			strResDir += _T("ip20\\");
+#else
+         strResDir += _T("ip20/");
+#endif
 		}
 		else if (strOem ==  _T("roip"))
 		{
+#ifndef __GNUC__s
 			strResDir += _T("roip\\");
+#else
+			strResDir += _T("roip/");
+#endif
 		}
+#ifndef __GNUC__s
 		strResDir += strRes + _T("\\");
+#else
+		strResDir += strRes + _T("/");
+#endif
 
 		strTempFile = strResDir + _T("login.htm"); 
+#ifndef __GNUC__
 		if (!CFile::GetStatus(strTempFile, status))
 		{	// use us http page if speical language not found
 			strResDir = strCurDir + _T("res\\us\\");
+#else
+      if (stat(strTempFile.c_str(),&status)) 
+		{	// use us http page if speical language not found
+			strResDir = strCurDir + _T("res/us/");
+#endif
 		}
 
 		listSrcName.AddTail(strResDir + _T("login.htm"));
@@ -356,7 +417,11 @@ BOOL _MergeFile(CString strCurDir, CStri
 		iOffset = 0;
 		for (i = 0; i < DSP_FILE_NUM; i ++)
 		{
+#ifndef __GNUC__
 			strFileName = strCurDir + _T("\\res\\") + _cDspFiles[i];
+#else
+			strFileName = strCurDir + _T("/res/") + _cDspFiles[i];
+#endif
 			if (!in.Open(strFileName, CFile::modeRead))
 			{
 				free(p);
@@ -410,7 +475,11 @@ BOOLEAN _WriteFile(CFile& out, CString s
 	memset(p, 0, iLength);
 	if (!in.Open(strFileName, CFile::modeRead))
 	{
+#ifndef __GNUC__
 		wprintf(_T("Unable to open %s\n"), strFileName);
+#else
+      printf(_T("Unable to open %s\n"), strFileName.c_str());
+#endif
 		return FALSE;
 	}
 	in.Read(p, iLength);
@@ -424,7 +493,11 @@ CString RenameBinary(CString strFileName
 {
 	CString strDstName, strCurDir, strSrcName;
 	CString strAllName, strSettingName, strDigitmapName, strPhonebookName, strPage0Name, strVersion, strRingtoneName, strHoldmusicName, strFontName, strUpgradeName;
+#ifndef __GNUC__
 	CFileStatus status;
+#else
+   struct stat status;
+#endif
 	CFile out;
 	CFile in;
 	int i, iLength;
@@ -433,22 +506,36 @@ CString RenameBinary(CString strFileName
 	CString strCmdLine, strSettingsFile, strDigitmapFile;
 
 	GetCurrentDirectory(128, szCurDir);
+#ifndef __GNUC__
 	strSrcName.Format(_T("%s\\%s"), szCurDir, strFileName);
 	strCurDir.Format(_T("%s\\"), szCurDir);
+#else
+   {
+      char buff[511];
+      sprintf(buff, "%s/%s", szCurDir, strFileName.c_str());
+      strSrcName = buff;
+      sprintf(buff, "%s/", szCurDir);
+      strCurDir = buff;
+   }
+#endif
 
 	strDstName = strCurDir + strVer + _T("_") + strCall + _T("_") + strRes + _T("_");
 	if (_IsOem(strOem))
 	{
 		strDstName += strOem + _T("_");
 	}
+
 	strVersion = _GetVersion(strCurDir);
 	strDstName += strVersion;
 
 	strAllName = strDstName;
 	strAllName += _T("_all.bin");
 	strDstName += _T(".bin");
-	
+#ifndef __GNUC__	
 	if (CFile::GetStatus(strDstName, status))
+#else
+   if (stat(strDstName.c_str(),&status))
+#endif
 	{
 		CFile::Remove(strDstName);
 	}
@@ -493,26 +580,49 @@ CString RenameBinary(CString strFileName
 	free(p);
 
 	// write settings_sip.dat or settings_iax2.dat
+#ifndef __GNUC__
 	strSettingsFile = strCurDir + _T("\\settings\\") + strVer + _T("_") + strCall + _T("_") + strRes;
+#else
+	strSettingsFile = strCurDir + _T("/settings/") + strVer + _T("_") + strCall + _T("_") + strRes;
+#endif
 	if (_IsOem(strOem))
 	{
 		strSettingsFile +=  _T("_") + strOem + _T(".txt");
 	}
 	else
 		strSettingsFile += _T(".txt");
-
+#ifndef __GNUC__	
 	if (!CFile::GetStatus(strSettingsFile, status))
-	{	// use default settings if speical .txt not found
+   {	// use default settings if speical .txt not found
 		strSettingsFile = strCurDir + _T("\\settings\\") + _T("default") + _T("_") + strCall + _T(".txt");
 	}
-
-	strCmdLine.Format(_T("..\\bin\\convert -c -f %s settings.dat"), strSettingsFile);
+   strCmdLine.Format(_T("..\\bin\\convert -c -f %s settings.dat"), strSettingsFile);
+#else
+   if (stat(strSettingsFile.c_str(),&status))
+	{	// use default settings if speical .txt not found
+		strSettingsFile = strCurDir + _T("/settings/") + _T("default") + _T("_") + strCall + _T(".txt");
+	}
+   {
+      char buff[511];
+      sprintf(buff, "../bin/convert -c -f %s settings.dat", strSettingsFile.c_str());
+      strCmdLine = buff;
+   }
+#endif
+#ifndef __GNUC__
 	_tsystem(strCmdLine);
+#else
+   printf("Executing comamnd: \"%s\"", strCmdLine.c_str());
+   if (!system(strCmdLine)) printf("Error executing comamnd: \"%s\"", strCmdLine.c_str());
+#endif
 
 	strSettingName = strCurDir + _T("settings.dat");
 	if (!in.Open(strSettingName, CFile::modeRead))
 	{
+#ifndef __GNUC__
 		wprintf(_T("Unable to open %s\n"), strSettingName);
+#else
+      printf(_T("Unable to open %s\n"), strSettingName.c_str());
+#endif
 		goto Error;
 	}
 	iLength = (int)in.GetLength();
@@ -537,13 +647,26 @@ CString RenameBinary(CString strFileName
 	free(p);
 
 	// write default settings
+#ifndef __GNUC__
 	strSettingsFile = strCurDir + _T("\\settings\\") + _T("default") + _T("_") + strCall + _T(".txt");
 	strCmdLine.Format(_T("..\\bin\\convert -c -f %s settings.dat"), strSettingsFile);
+#else
+   strSettingsFile = strCurDir + _T("/settings/") + _T("default") + _T("_") + strCall + _T(".txt");
+   {
+      char buff[511];
+      sprintf(buff, "../bin/convert -c -f %s settings.dat", strSettingsFile.c_str());
+      strCmdLine = buff;
+   }
+#endif
 	_tsystem(strCmdLine);
 	strSettingName = strCurDir + _T("settings.dat");
 	if (!in.Open(strSettingName, CFile::modeRead))
 	{
+#ifndef __GNUC__
 		wprintf(_T("Unable to open %s\n"), strSettingName);
+#else
+      printf(_T("Unable to open %s\n"), strSettingName.c_str());
+#endif
 		goto Error;
 	}
 	iLength = (int)in.GetLength();
@@ -568,21 +691,33 @@ CString RenameBinary(CString strFileName
 	free(p);
 
 	// Write phonebook.dat
+#ifndef __GNUC__
 	strPhonebookName = strCurDir + _T("\\res\\") + _T("phonebook.dat");
+#else
+	strPhonebookName = strCurDir + _T("/res/") + _T("phonebook.dat");
+#endif
 	if (!_WriteFile(out, strPhonebookName, PHONEBOOK_FILE_SIZE))
 	{
 		goto Error;
 	}
 
 	// write ringtone.dat
+#ifndef __GNUC__
 	strRingtoneName = strCurDir + _T("\\res\\") + _T("ringtone.dat");
+#else
+	strRingtoneName = strCurDir + _T("/res/") + _T("ringtone.dat");
+#endif
 	if (!_WriteFile(out, strRingtoneName, SYSTEM_RINGTONE_PAGE_NUM * FILE_FULL_PAGE_SIZE))
 	{
 		goto Error;
 	}
 
 	// write ringtone_hold.dat
+#ifndef __GNUC__
 	strHoldmusicName = strCurDir + _T("\\res\\") + _T("holdmusic.dat");
+#else
+	strHoldmusicName = strCurDir + _T("/res/") + _T("holdmusic.dat");
+#endif
 	if (!_WriteFile(out, strHoldmusicName, SYSTEM_HOLDMUSIC_PAGE_NUM * FILE_FULL_PAGE_SIZE))
 	{
 		goto Error;
@@ -591,11 +726,19 @@ CString RenameBinary(CString strFileName
 	// write font.dat or ivr_xx.dat
 	if (strOem == _T("ivr"))
 	{
+#ifndef __GNUC__
 		strFontName = strCurDir + _T("\\res\\") + _T("ivr_") + strRes + _T(".dat");
+#else
+	   strFontName = strCurDir + _T("/res/") + _T("ivr_") + strRes + _T(".dat");
+#endif
 	}
 	else
 	{
+#ifndef __GNUC__
 		strFontName = strCurDir + _T("\\res\\") + _T("font.dat");
+#else
+		strFontName = strCurDir + _T("/res/") + _T("font.dat");
+#endif
 	}
 	if (!_WriteFile(out, strFontName, SYSTEM_FONT_PAGE_NUM * FILE_FULL_PAGE_SIZE))
 	{
@@ -637,12 +780,24 @@ void ReportCodeSize()
 	}
 
 // _CODE7                 00008000    000052FC =       21244. bytes (REL,CON)
+#ifndef __GNUC__
 	while (f.ReadString(str))
 	{
+#else
+   while (!f.eof())
+   {
+      CString str = "";
+      std::getline(f, str);
+      if (*str.rbegin() == '\r') str.erase(str.length() - 1);
+#endif
 		if ((str.Left(5) == _T("_CODE")) && (str.Right(9) == _T("(REL,CON)")))
 		{
 			bFound = false;
-			for (pos = list.GetHeadPosition(); pos != NULL;)
+#ifndef __GNUC__
+         for (pos = list.GetHeadPosition(); pos != NULL;)
+#else
+         for (pos=list.begin(); pos != list.end(); ++pos)
+#endif
 			{
 				if (str == list.GetNext(pos))
 				{
@@ -658,7 +813,11 @@ void ReportCodeSize()
 	}
 
 	iTotal = 0;
+#ifndef __GNUC__
 	for (pos = list.GetHeadPosition(); pos != NULL;)
+#else
+   for (pos=list.begin(); pos != list.end(); ++pos)
+#endif
 	{
 		str = list.GetNext(pos);
 //		wprintf(_T("%s\n"), str);
@@ -679,7 +838,9 @@ void ZipBin(CString strDstName)
 	strZipName = strDstName.Left(strDstName.GetLength() - 3)  + _T("zip");
 
 	// zip .bin file
+#ifndef __GNUC__
 	strCmdLine.Format(_T("..\\bin\\7z a %s %s"), strZipName, strDstName);
 	_tsystem(strCmdLine);
+#endif
 }
 
diff -rupN '--exclude=.*' ../SDCC/tool/relink/Makefile ./tool/relink/Makefile
--- ../SDCC/tool/relink/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/relink/Makefile	2013-02-22 11:06:45.000000000 +0200
@@ -0,0 +1,14 @@
+CC=g++
+CFLAGS=-I.
+OBJ = relink.o ../common/cmd_default.o 
+
+all: relink
+
+%.o: %.cpp
+	$(CC) -c -o $@ $< $(CFLAGS)
+
+relink: $(OBJ)
+	g++ -o $@ $^ $(CFLAGS)
+
+clean:
+	rm -f *.o *~ ../common/*.o relink
diff -rupN '--exclude=.*' ../SDCC/tool/relink/StdAfx.h ./tool/relink/StdAfx.h
--- ../SDCC/tool/relink/StdAfx.h	2012-06-29 12:33:32.000000000 +0300
+++ ./tool/relink/StdAfx.h	2013-02-22 09:59:12.000000000 +0200
@@ -6,6 +6,8 @@
 #if !defined(AFX_STDAFX_H__A66AE15E_51B5_47E3_A910_9BC64B0E928E__INCLUDED_)
 #define AFX_STDAFX_H__A66AE15E_51B5_47E3_A910_9BC64B0E928E__INCLUDED_
 
+#ifndef __GNUC__
+
 #if _MSC_VER > 1000
 #pragma once
 #endif // _MSC_VER > 1000
@@ -30,7 +32,16 @@
 
 // TODO: reference additional headers your program requires here
 
+#include "..\\common\\common.h"
+#include "..\\..\\include\\version.h"
+
 //{{AFX_INSERT_LOCATION}}
 // Microsoft Visual C++ will insert additional declarations immediately before the previous line.
 
+#else // __GNUC__
+#include "../common/mfc2std.h"
+#include "../common/common.h"
+#include "../../include/version.h"
+#endif
+
 #endif // !defined(AFX_STDAFX_H__A66AE15E_51B5_47E3_A910_9BC64B0E928E__INCLUDED_)
diff -rupN '--exclude=.*' ../SDCC/tool/relink/relink.cpp ./tool/relink/relink.cpp
--- ../SDCC/tool/relink/relink.cpp	2013-02-16 20:13:12.000000000 +0200
+++ ./tool/relink/relink.cpp	2013-02-22 09:59:34.000000000 +0200
@@ -1,11 +1,7 @@
 // relink.cpp : Defines the entry point for the console application.
 //
 
-#include "stdafx.h"
-
-#include "..\\common\\common.h"
-
-#include "..\\..\\include\\version.h"
+#include "StdAfx.h"
 
 #ifdef _DEBUG
 #define new DEBUG_NEW
@@ -17,7 +13,7 @@ void _RewriteLinkFile(CString strFileNam
 
 /////////////////////////////////////////////////////////////////////////////
 // The one and only application object
-
+#ifndef __GNUC__
 CWinApp theApp;
 
 using namespace std;
@@ -45,11 +41,22 @@ int _tmain(int argc, TCHAR* argv[], TCHA
 
 	return nRetCode;
 }
+#else
+int main(int argc, char *argv[])
+{
+	int nRetCode = 0;
+
+		CString strFileName, strVer, strCall, strRes, strOem;
+		printf("Palmmicro AR1688 relink utility %d.%d%d.%d%d%d\n", SOFT_VER_HIGH, SOFT_VER_LOW, SOFT_VER_BUILD, ENG_BUILD_HIGH, ENG_BUILD_MID, ENG_BUILD_LOW);
+		CommandLineDefault(argc, argv, strFileName, strVer, strCall, strRes, strOem);
+		_RewriteLinkFile(strFileName, strVer, strCall, strRes, strOem);
+	return nRetCode;
+}
+#endif
 
 BOOL _ProcessString(CString & str, CString strVer, CString strCall, CString strRes, CString strOem)
 {
 	CString strNew;
-
 	if (str.Left(7) != _T("..\\lib\\"))		return FALSE;
 
 	if (str.Right(7) == _T("arp.rel"))
@@ -203,28 +210,50 @@ void _RewriteLinkFile(CString strFileNam
 	BOOL bChanged;
 
 	GetCurrentDirectory(128, szCurDir);
+#ifndef __GNUC__
 	strName.Format(_T("%s\\%s"), szCurDir, strFileName);
-
+#else
+   {
+      char buff[511];
+	   sprintf(buff, "%s/%s", szCurDir, strFileName.c_str());
+      strName = buff;
+   }
+#endif
 	strVer.MakeLower();
 	strCall.MakeLower();
 	strRes.MakeLower();
 	strOem.MakeLower();
-
 	// read first
 	if (!file.Open(strName, CFile::modeRead|CFile::typeText))
 	{
+#ifndef __GNUC__
 		wprintf(_T("Can not open file %s for read\n"), strName);
+#else
+      printf(_T("Can not open file %s for read\n"), strName.c_str());
+#endif
 		return;
 	}
+#ifndef __GNUC__
 	while (file.ReadString(str))
 	{
+#else
+   while (!file.eof())
+   {
+      CString str = "";
+      std::getline(file, str);
+      if (*str.rbegin() == '\r') str.erase(str.length() - 1);
+#endif
 		list.AddTail(str);
 	}
 	file.Close();
 
 	// process
 	bChanged = FALSE;
+#ifndef __GNUC__
 	for (pos = list.GetHeadPosition(); pos != NULL;)
+#else
+   for (pos=list.begin(); pos != list.end(); ++pos)
+#endif
 	{
 		old = pos;
 		str = list.GetNext(pos);
@@ -234,15 +263,22 @@ void _RewriteLinkFile(CString strFileNam
 			bChanged = TRUE;
 		}
 	}
-
+#ifndef __GNUC__
 	if (!bChanged)	return;		// no need to rewrite
-
 	// write
 	file.Open(strName, CFile::modeCreate|CFile::modeWrite|CFile::typeText);
 	while (!list.IsEmpty())
 	{
 		str = list.RemoveHead();
 		str += _T("\n");
+#else
+	file.Open(strName.Left(strName.length()-4)+"_unix.lnk",std::fstream::out|std::fstream::trunc);
+   for(pos=list.begin(); pos!=list.end(); pos++)
+   {
+      str = *pos;
+      str.FindAndReplace("\\","/");
+      str += _T("\r\n");
+#endif
 		file.WriteString(str);
 	}
 	file.Close();
diff -rupN '--exclude=.*' ../SDCC/tool/reversion/Makefile ./tool/reversion/Makefile
--- ../SDCC/tool/reversion/Makefile	1970-01-01 02:00:00.000000000 +0200
+++ ./tool/reversion/Makefile	2013-02-22 11:05:57.000000000 +0200
@@ -0,0 +1,14 @@
+CC=g++
+CFLAGS=-I.
+OBJ = reversion.o ../common/cmd_default.o 
+
+all: reversion
+
+%.o: %.cpp
+	$(CC) -c -o $@ $< $(CFLAGS)
+
+reversion: $(OBJ)
+	g++ -o $@ $^ $(CFLAGS)
+
+clean:
+	rm -f *.o *~ ../common/*.o reversion
diff -rupN '--exclude=.*' ../SDCC/tool/reversion/StdAfx.h ./tool/reversion/StdAfx.h
--- ../SDCC/tool/reversion/StdAfx.h	2012-06-29 12:33:32.000000000 +0300
+++ ./tool/reversion/StdAfx.h	2013-02-22 09:55:55.000000000 +0200
@@ -6,6 +6,8 @@
 #if !defined(AFX_STDAFX_H__7B115433_523B_48A4_BB03_9769F39B7FAB__INCLUDED_)
 #define AFX_STDAFX_H__7B115433_523B_48A4_BB03_9769F39B7FAB__INCLUDED_
 
+#ifndef __GNUC__
+
 #if _MSC_VER > 1000
 #pragma once
 #endif // _MSC_VER > 1000
@@ -29,8 +31,15 @@
 #include <iostream>
 
 // TODO: reference additional headers your program requires here
+#include "..\\common\\common.h"
 
 //{{AFX_INSERT_LOCATION}}
 // Microsoft Visual C++ will insert additional declarations immediately before the previous line.
 
+#else // __GNUC__
+#include "../common/mfc2std.h"
+#include "../common/common.h"
+#endif
+
 #endif // !defined(AFX_STDAFX_H__7B115433_523B_48A4_BB03_9769F39B7FAB__INCLUDED_)
+
diff -rupN '--exclude=.*' ../SDCC/tool/reversion/reversion.cpp ./tool/reversion/reversion.cpp
--- ../SDCC/tool/reversion/reversion.cpp	2012-06-29 12:33:32.000000000 +0300
+++ ./tool/reversion/reversion.cpp	2013-02-22 09:54:59.000000000 +0200
@@ -1,9 +1,7 @@
 // reversion.cpp : Defines the entry point for the console application.
 //
 
-#include "stdafx.h"
-
-#include "..\\common\\common.h"
+#include "StdAfx.h"
 
 #ifdef _DEBUG
 #define new DEBUG_NEW
@@ -15,7 +13,7 @@ void _RewriteVersion(CString strFileName
 
 /////////////////////////////////////////////////////////////////////////////
 // The one and only application object
-
+#ifndef __GNUC__
 CWinApp theApp;
 
 using namespace std;
@@ -42,7 +40,19 @@ int _tmain(int argc, TCHAR* argv[], TCHA
 
 	return nRetCode;
 }
+#else
+int main(int argc, char *argv[])
+{
+	int nRetCode = 0;
 
+		CString strFileName, strVer, strCall, strRes, strOem;
+		
+		printf("Palmmicro reversion utility 0.55.011\n");
+		CommandLineDefault(argc, argv, strFileName, strVer, strCall, strRes, strOem);
+		_RewriteVersion(strFileName, strVer, strCall, strRes, strOem);
+	return nRetCode;
+}
+#endif
 BOOL _MatchString(CString str, CString strCmp)
 {
 	int iLen, iVal;
@@ -104,8 +114,15 @@ void _RewriteVersion(CString strFileName
 	BOOL bChanged;
 
 	GetCurrentDirectory(128, szCurDir);
+#ifndef __GNUC__
 	strName.Format(_T("%s\\%s"), szCurDir, strFileName);
-
+#else
+   {
+      char buff[511];
+      sprintf(buff, "%s/%s", szCurDir, strFileName.c_str());
+      strName = buff;
+   }
+#endif
 	strCmpVer = _T("VER_") + strVer;
 	strCmpVer.MakeUpper();
 
@@ -121,18 +138,34 @@ void _RewriteVersion(CString strFileName
 	// read first
 	if (!file.Open(strName, CFile::modeRead|CFile::typeText))
 	{
+#ifndef __GNUC__
 		wprintf(_T("Can not open file %s for read\n"), strName);
+#else
+      printf(_T("Can not open file %s for read\n"), strName.c_str());
+#endif
 		return;
 	}
+#ifndef __GNUC__
 	while (file.ReadString(str))
 	{
+#else
+   while (!file.eof())
+   {
+      CString str = "";
+      std::getline(file, str);
+      if (*str.rbegin() == '\r') str.erase(str.length() - 1);
+#endif
 		list.AddTail(str);
 	}
 	file.Close();
 
 	// process
 	bChanged = FALSE;
+#ifndef __GNUC__
 	for (pos = list.GetHeadPosition(); pos != NULL;)
+#else
+   for (pos=list.begin(); pos != list.end(); ++pos)
+#endif
 	{
 		old = pos;
 		str = list.GetNext(pos);
@@ -151,10 +184,17 @@ void _RewriteVersion(CString strFileName
 
 	// write
 	file.Open(strName, CFile::modeCreate|CFile::modeWrite|CFile::typeText);
+#ifndef __GNUC__
 	while (!list.IsEmpty())
 	{
 		str = list.RemoveHead();
 		str += _T("\n");
+#else
+   for(pos=list.begin(); pos!=list.end(); pos++)
+   {
+      str = *pos;
+      str += _T("\r\n");
+#endif
 		file.WriteString(str);
 	}
 	file.Close();
